### ISSUE TITLE
Interaction with shapes

### ISSUE THREAD

**Issue by VeyDlin** [mono/SkiaSharp#2330]
I'm trying to determine which shapes are located by coordinates (mouse click).

I realized that skia does not provide such an API and I need to implement it myself, but I have no ideas on this.

Everything is complicated by the fact that I need to understand exactly that the click occurred on the figure, and not its square hitbox, for example, I can click inside the text with the number "0" and this should not be considered a hit, there are also rectangles with rounded edges, complex curves, different shapes, and there are thousands of them on the canvas

I will be grateful for any advice because right now I don't even know how to Google for a solution to my problem

Я пытаюсь определить какие фигуры находятся по координатам (клик мышки).

Я понял, что skia не предоставляет такой API и мне надо реализовывать это самому, но у меня нет никаких идей на этот счет.

Все осложняется тем, что мне необходимо точно понимать, что клик произошел именно на фигуру, а не ее квадратный хитбокс, например я могу кликнуть внутри текста с цифрой "0" и это не должно считаться за попадание, так же есть прямоугольники с закругленными краями, сложные кривые, разные фигуры...

Я буду благодарен за любой совет потому что сейчас я даже не знаю, как мне искать в гугл решение моей проблемы

**Comment by themcoo** [comment]
Skia is a rendering library, what you are asking for is completely out of scope.

**Comment by VeyDlin** [comment]
> Skia is a rendering library, what you are asking for is completely out of scope.

Yes, I understand that, I just wanted to find out what ways there are in general, 

I could create a mathematical model of each figure I draw and count its collision, 

or I could add its own canvas to each drawn figure and draw it there again and get a pixel at the right coordinates (if there is such a function in skia at all), then I will get thousands of canvases, or tens of thousands, and maybe it won't even work too slowly,

I could also create a polygon around pixels, for this I also need to somehow get a drawn shape separately from the general canvas, is this possible in skia?

But it still seems to me that there should be an easier way, if skia returned at least some information about the shape that it just rendered, it would be easier, I think

**Comment by willtn05** [comment]
you could draw the shapes to a new surface, pass a unique color for each shape, then call `hittestsurface.ReadPixels(..., mousePos)` and check if the pixel color is a match?

**Comment by utenma** [comment]
this is just speculation at the moment but representing the UI as a DOM-like tree of System.Drawing.Drawing2D objects then mapping the DOM-like tree to skia drawing methods should be one approach

Then on pointer events use

[Rectangle.Contains](https://learn.microsoft.com/en-us/dotnet/api/system.drawing.rectangle.contains?view=net-8.0)
[GraphicsPath.IsVisible](https://learn.microsoft.com/en-us/dotnet/api/system.drawing.drawing2d.graphicspath.isvisible?view=net-8.0&redirectedfrom=MSDN#overloads)

For prioritizing elements of top of others, the DOM-like tree need z-index

@VeyDlin issue should be closed and instead a discussion thread in discussions as it's outside skia codebase

**Comment by wieslawsoltes** [comment]
You can try https://github.com/wieslawsoltes/Draw2D

**Comment by Aniganesh** [comment]
Hi @wieslawsoltes! I opened your library and it seems the github pages deployment has failed.
